### PR TITLE
New API Design Doc

### PR DESCRIPTION
**Pull request by wangkuiyi**

### CODE REVIEW

File: doc/design/api.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Evaluator = {Model*, activations}
+- forward
+- test
+
+GradientMachine = {Model*, gradients}
+- backward
+
+Optimizer = {Model*, Evaluator*, GradientMachine*}
+- train
+- update
```
Comment (helinwang) [inline doc/design/api.md]: I think `update` don't need to be exposed publicly (only used by train).

File: doc/design/api.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+- checkpoint
```
Comment (helinwang) [inline doc/design/api.md]: Does this checkpoint method mean save everything (model topology, parameters, optimizer states)?

I think checkpoint should be called internally from train, and not exposed.
Instead, expose
```
serialized_state = sgd.save() # save only internal state of optimizer
sgd = paddle.optimizer.SGD(serialized_state) # create SGD optimizer from serialized_state
```
In this way everything is orthogonal (before `optimizer.checkpoint` includes `model.save`), user will have option to save and load whatever they want (typically when they manually implement `train`). By doing:
```
# save
model.save(path)
sgd_state = sgd.save()

# load
model.load(path)
sgd = paddle.optimizer.SGD(sgd_state)
```

File: doc/design/api.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
+# Design Doc: PaddlePaddle API
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
```
Comment (helinwang) [inline doc/design/api.md]: One comment for future when we will add bindings for more languages (or even now, if people use python to create topology and train with c++)

Now model parameter is initialized during model creation: `model = paddle.model.create()`.
`model.save()` will always save both `parameter` and `topology`, and `model.load()` will load both as well.
We may need a method `model.init()` to initialize model parameters again.

The use case is: When user create topology from python (because it's easier), and use it in C++ (or golang in future). User will like to have the option to init parameter again in order to start training from somewhere else in the parameter space.

From the use case above, we can see that topology and parameter is not tightly coupled. Maybe we can consider instead expose
```
save_topology()
save_parameter()
load_topology()
load_parameter()
```

Comment (wangkuiyi) [inline doc/design/api.md]: I think "creating topology using Python and run training using C++/Go" is not a common case, but due to that Tensorflow doesn't provide a language binding that works both good for specifying topology and for training. In other words, I don't see a chance that we need to load only topology, or only parameters, but not both.

File: doc/design/api.md (hunk 1 above)
Comment (wangkuiyi) [inline doc/design/api.md]: If we are going to use *composition* other than *derivation*, then I would agree that we should have Optimizer, which has the `train` method, and Updater, which has the `update` method.  But as Python supports derivation, we can create AdamOptimizer basing on SGDOptimizer by simply overriding `Optimizer::update`.

File: doc/design/api.md (hunk 2 above)
Comment (wangkuiyi) [inline doc/design/api.md]: Actually, i have no much idea about checkpointing. We might consult other team member for what checkpointing needs to do.

File: doc/design/api.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+Evaluator = {Model*, activations}
+- forward
+- test
+
+GradientMachine = {Evaluator*, gradients}
+- backward
+
+Optimizer = {Model*, Evaluator*, GradientMachine*}
```
Comment (reyoung) [inline doc/design/api.md]: Optimizer = {GradientMachine*}

GradientMacine -> Evaluator -> Model

File: doc/design/api.md (hunk 3 above)
Comment (reyoung) [inline doc/design/api.md]: maybe `serialize`, `deserialize` is better.

Comment (jacquesqiao) [inline doc/design/api.md]: should we add a concept of Cost, there are several kind of cost, and different cost may compose to a new kind of Cost.

File: doc/design/api.md [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
+
+GradientMachine = {Evaluator*, gradients}
+- backward
```
Comment (reyoung) [inline doc/design/api.md]: Maybe gradient_machine's API could be

```python
result = gradient_machine.forward_backward(..., on_parameter_backwarded=None)  # basically same arguments in `Evaluator.forward`
```

The details show below.

* on_parameter_backwarded will be invoked when the parameter's gradient is calculated immediately. This callback is very useful for network communication. We could send some parameter's gradient while other gradients are calculating.
* Combine forward and backward into one method, because it is not necessarily two stage, especially for multi-threading calculation.
   * The multi-thread backward could not wait for all thread forwarding complete. Each thread could do forward-backward individually. So combine `forward-backward` method together, could let we have a chance to optimize multi-thread training.
   * Not every forwarding can be backward. We must record all activations for backward. But for model inference, forward only need get the final output, intermediate activations could be dropped for saving memory.

File: doc/design/api.md [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+Optimizer = {GradientMachine*}
+- train(cost, ...)
```
Comment (reyoung) [inline doc/design/api.md]:
```python
def reader():
    yield {'pixel': pixels, 'label': labels}  # return a data batch.

# Observe callback is used for plotting or logging the training process.
# The type of event parameter could be various. The intermediate result for training is in event instance.

def callback(event):
     if isinstance(event, FinishTrainOneBatch):
        print event.pass_id, event.batch_id, "Cost = ", event.cost, "Error Rate = ", event.metric[0]
        print "output layer's output is ", event.activation['output']
     else:
        pass

optimizer.train(data_reader=reader,
                cost=CrossEntropy(input=model.topology.output_layer,  # the network's output layer.
                                  label=DataReader("label")),  # Label is get from data_reader's 'label' field.
                metric=[ErrorRateMetric(input=model.topology.output_layer, label=DataReader("label"))], # same logic above
                observe_callback=callback
)
```

File: doc/design/api.md [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
```
Comment (reyoung) [inline doc/design/api.md]: Even more aggressively, Is `Evaluator` a necessary concept for Paddle?

It seems if we think the activation should be the result of forwarding a neural network, `Evaluator` should store nothing. The global `forward` method is enough? or `forward` method should be a part of neural network model?

File: doc/design/api.md (hunk 5 above)
Comment (hedaoyuan) [inline doc/design/api.md]: 嗯，我们如果提供的是一个封装了`多CPU/GPU/多机`这样的梯度计算接口，那么这里应该是一个forward_backward更适合，可以隐藏通信这些信息。

Comment (reyoung) [inline doc/design/api.md]: 恩，另外在我的最激进的想象里(可以忽略，这个想法没有仔细想过性能问题)，这个forward_backward可能也是纯粹实现在Python里面。

例如，对于同时使用12个CPU和4个GPU的gradient_machine。实现起来的伪代码可能是

```python
class GradientMachine():
    def __init__(self, model):
        self.origin_model = model
        self.cpu_models = []
        self.gpu_models = []
        for i in xrange(12):
            self.cpu_models.append(paddle.model.shared_model(model))

        for i in xrange(4):
            self.gpu_models.append(paddle.model.copy_model(model, device=i))

    def forward_backward(data):
        datas = data.split(len(self.cpu_models) + len(self.gpu_models))
        result = paddle.thread_pool.map(lambda model, data: model.forward_backward(model, data), 
                                        zip([self.cpu_models, self.gpu_models], datas))
        gpu_gradient = paddle.gpu.merge_gradient(self.gpu_models)
        cpu_gradient = self.origin_model.gradient 
        cpu_gradient += gpu_gradient
        result.gradient = cpu_gradient
        return result
```

File: doc/design/api.md (hunk 7 above)
Comment (hedaoyuan) [inline doc/design/api.md]: 这里用户接口层面的Model增加`device`属性，可能会给使用者带来一些困惑。不一定所有的人都理解`But, when Model in one GPU card, we cannot forward it by another GPU card.`，那么当使用者构造了一个Model(device=0)的model，但在device=1上执行失败，就会比较费解。

Comment (helinwang) [inline doc/design/api.md]: @reyoung Good comments! For the "add `begin` and `end`" part, I don't think `begin` is necessary? Because the dependency is fully specified when we want to calculate `end`.

Comment (helinwang) [inline doc/design/api.md]: For the "Is `Evaluator` a necessary concept for Paddle?". I think we need some object for caching intermediate tensors. From what I understand (asked this question in yesterday's meeting), paddle don't release intermediate tensors after a forward pass even only for inference. I think that is reasonable. Since `malloc` and `free` takes time. If we don't release intermediate tensors, but reusing them by overriding, we can save forward time. (at least, release or not could be tunable argument).
So if we don't release, we need an object to save the states(all intermediate tensors). `Evaluator` could be such an object.

File: doc/design/api.md (hunk 3 above)
Comment (wangkuiyi) [inline doc/design/api.md]: 如果考虑“a model's parameter could be reused by another model.”，那么需要对reuse的过程有所理解。可否写一段伪码试试，以便理解这个问题？

我试了一下：

```
in = layer.data(...)
int = layer.fc(in, ..., parameter_name="proj")
out = layer.softmax(int, ...)
m = model.create(out)
m.parameter("proj").deserialize(stream)
```

显然有几个问题：

1. stream 从哪儿来？
1. model的其他参数从哪儿来？

如果把这两个问题及其延伸问题都考虑清楚，并且描述清楚，那么这个文档又要增长多长？我们需要多少时间来实现？对第一期目标“让目前八章book更简短”有什么帮助？

目前我们讨论的是个框架，就像Unix一开始并没有现在这么强大，但是核心思路奠定了未来可以支持更多的功能，那就是一个好的设计。

想到新的功能不难，知道可以不用做什么才是真功夫。基于目前的设计，复用参数的功能是不是日后可以再加上去的？

**Comment by wangkuiyi** [comment]
@jacquesqiao https://github.com/PaddlePaddle/Paddle/pull/1297#discussion_r100474526 里提出的是个好问题。大家怎么看？

File: doc/design/api.md (hunk 7 above)
Comment (helinwang) [inline doc/design/api.md]: I don't totally understand why "Model should be bound to a device(a GPU card, or all CPUs)". I think parameters should be place by paddle automatically. Paddle should try it's best to place all parameters on GPU. The only case it fails to do so is when a layer don't have GPU implementation. So parameter will be allocated on CPU memory. And input tensor for that layer will be copied from GPU to CPU. output tensor will be copied from CPU to GPU.

It will be same for multi-GPU case, model is replicated n times where n is number of GPUs. So when initializing `Evaluator` and `Gradient Machine`, constructor parameters need to include GPU device ids as a list, `len(list)` will be how many model replicas we have.

Comment (wangkuiyi) [inline doc/design/api.md]: 我也觉得model里不应该指定device。我的考虑是现代分布式计算里，在哪个设备上执行某个操作应该是调度系统（Linux kernel、Kubernetes scheduler、PaddlePaddle framework）自动决定的，而不是用户指定的。唯有沿着这个思路，我们才能做出一个真正“易用”的DL系统，和目前很多学术界做出来的系统有所区别。

Comment (wangkuiyi) [inline doc/design/api.md]: @jacquesqiao #1297 (comment) 里提出的是个好问题。大家怎么看？

Comment (helinwang) [inline doc/design/api.md]: 龙飞说的"should we add a concept of Cost, there are several kind of cost, and different cost may compose to a new kind of Cost."很有道理。以下是我的理解：

我觉得类型上，cost应该就是一个layer。任何一个输出是scalar的layer都可以作为cost。因为我们做backprop的时候，是把那个layer的输出设成1来做的，这里需要且仅需要那一个layer的输出是一个scalar。

使用方法／概念上，可以把cost layer放在模型外
```
forward(cost=cross_entropy(y,y_)) // or test(cross_entropy(y,y_))
backward(cross_entropy(y,y_))
```
或者模型内
```
forward(cross_entropy_loss) // 完全不需要test(cross_entropy(y,y_))，因为都是layer
backward(cross_entropy_loss)
```
从概念简单来看是后者好，但是从模型简单，可扩展（比如说load一个topology，换cost）来看，其实是前者好。

File: doc/design/api.md [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
```
Comment (wangkuiyi) [inline doc/design/api.md]: @reyoung 的https://github.com/PaddlePaddle/Paddle/pull/1297/files#r100475149 里提出的问题，我的看法是：

在 trianing 的时候forward肯定得保存activations为class member，因为backward需要使用。

即便在 serving 的时候，我们也不能假设只需要保存输出层的activations —— 万一用户想visualize中间某层的输出呢？

File: doc/design/api.md (hunk 7 above)
Comment (helinwang) [inline doc/design/api.md]: 我觉得model可以是这样
model
|-topology
|-parameters
   |-parameter1
      |-parameter1-on-GPU1
      |-parameter1-on-GPU2
   |-parameter2
   |-...
所以不需要给model指定device，只用Evaluator和Gradient Machine跟model说哪个parameter对应的内存分配在哪里了。

Comment (wangkuiyi) [inline doc/design/api.md]: 我也理解cost是layer。

Comment (wangkuiyi) [inline doc/design/api.md]: 沿着 @helinwang 说的，进一步考虑：我认为一个model里所有parameters都放在一个gpu的显卡里就好了（或者在没有gpu的时候都放在一台机器的内存里）。至少项目的第一阶段如此。一个复杂的设计，很容易让我们丧失对项目的把控，即便能带来性能的优化，也往往得不偿失。

File: doc/design/api.md (hunk 5 above)
Comment (helinwang) [inline doc/design/api.md]: 我觉得没有forward_backward的时候，也可以不等待所有线程都完成再backward。这个要看train怎么实现了。每个线程里forward之后可以立刻backward呀。
比如：
```
// thread
forward()
backward()
sync_all()
```
你说的是
```
// thread
forward()
sync_all()
backward()
sync_all()
```

File: doc/design/api.md (hunk 7 above)
Comment (reyoung) [inline doc/design/api.md]: cost是一个Layer，或者更简单的，可以是Python端定义的一个函数。

与此同时，cost是目前forward的一个参数。也就是这个Layer是在forward的时候，attach到模型(Model)上的。

Comment (reyoung) [inline doc/design/api.md]: > 我认为一个model里所有parameters都放在一个gpu的显卡里就好了（或者在没有gpu的时候都放在一台机器的内存里）

同意。

可以不用太在意那个`aggressively`的想法。如果我们对整体概念没有巨大分歧的话，是否可以先讲design doc merge掉，我们开始撸起袖子写接口了?

File: doc/design/api.md (hunk 8 above)
Comment (reyoung) [inline doc/design/api.md]: 1、result.activation保存了所有层的activation。所以，backward和用户需要可视化的时候，都可以使用。
2、将activation返回，也是为了让forward这个函数可以变得『无状态』，也就是一个纯粹的函数。这样做的好处是在某些类型的任务中(例如机器翻译，这个是前8章的内容)，预测会多次调用forward函数，并且forward函数需要在神经网络的中间开始调用(也就是神经网络的Activations需要被Get/Set)。这如果将activation保存到 data member里面，会让这个过程变得略微困难一点。

**Comment by reyoung** [comment]
有一个问题需要周知一下，就是神经网络的预测，不一定是只调用一次forward就搞定了。很多模型中，会对forward的结果，神经网络的中间结点进行修正。以PaddlePaddle/book中的机器翻译举例如下:

image


机器翻译训练的是一个生成模型，他使用一个生成式的RNN作为decoder。这个docoder实际上会连续的调用一个以目标语言词表为宽度的softmax函数。生成目标语言里面每个单词的概率。训练的时候，使用贪心法训练，每一步都取这个目标语言词表里最大概率的值，作为这个RNN下一个时间步的输入。即上图。

但是在预测的时候，每一个时间步，不会仅仅再每一步取最大的概率值(不会有ArgMax那个节点)，而是采取一个更优化的树搜索策略(类似于剪枝的BFS, BeamSearch)。因为贪心法肯定不是最优解。

于是，我们会在ArgMax那个节点，分出N多岔路。并且会在每一个岔路里面，只再forward ArgMax右侧的部分节点。

这种需求，要求我们的forward函数是，可以从网络的中间部分开始forward，并且forward一部分网络。

**Comment by wangkuiyi** [comment]
OK. I am following @reyoung 's suggestion and merging this PR. But there are more to be discussed as in https://github.com/PaddlePaddle/Paddle/issues/1315. I am opening a new PR for that.
